sdk/lib/opentelemetry/sdk/trace/span.rb

```diff
@@ -332,12 +332,12 @@ def trim_links(links, link_count_limit, link_attribute_count_limit) # rubocop:di
           return nil if links.nil?
 
           if links.size <= link_count_limit &&
-             links.all? { |link| link.attributes.size <= link_attribute_count_limit && Internal.valid_attributes?(name, 'link', link.attributes) }
+             links.all? { |link| link.span_context.valid? && link.attributes.size <= link_attribute_count_limit && Internal.valid_attributes?(name, 'link', link.attributes) }
             return links.frozen? ? links : links.clone.freeze
           end
 
           # Slow path: trim attributes for each Link.
-          links.last(link_count_limit).map! do |link|
+          links.select { |link| link.span_context.valid? }.last(link_count_limit).map! do |link|
```

Comment:
I can't see a clean way to only allocate one array here, unfortunately. 🤔 I guess we could do the equivalent of `excess.times { links.shift }` after the `links.select`.